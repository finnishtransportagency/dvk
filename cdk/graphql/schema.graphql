schema {
  query: Query
  mutation: Mutation
}

type Query {
  fairwayCards(status: [Status!]): [FairwayCard!]!
  fairwayCard(id: String!, version: String!): FairwayCard
  fairwayCardVersion(id: String!, version: String!): FairwayCard
  fairwayCardLatest(id: String!): FairwayCard
  fairwayCardPublic(id: String!): FairwayCard
  fairwayCardPreview(id: String!, version: String!): FairwayCard
  safetyEquipmentFaults: [SafetyEquipmentFault!]!
  marineWarnings: [MarineWarning!]!
  currentUser: CurrentUser
  pilotPlaces: [PilotPlace!]!
  harbors: [Harbor!]!
  harbor(id: String!, version: String!): Harbor
  harborVersion(id: String!, version: String!): Harbor
  harborLatest(id: String!): Harbor
  harborPublic(id: String!): Harbor
  harborPreview(id: String!, version: String!): Harbor
  fairways: [Fairway!]!
  fairwayCardsAndHarbors: [FairwayCardOrHarbor!]!
  mareographs: [Mareograph!]!
}

type Mutation {
  saveFairwayCard(card: FairwayCardInput!): FairwayCard
  saveHarbor(harbor: HarborInput!): Harbor
  uploadPicture(picture: PictureUploadInput!): Boolean
}

input PictureUploadInput {
  cardId: String!
  id: String!
  base64Data: String!
  contentType: String!
}

enum Operation {
  CREATE
  UPDATE
}

enum OperationError {
  CARD_ID_MISSING
  CARD_ALREADY_EXIST
  CARD_NOT_EXIST
  HARBOR_ID_MISSING
  HARBOR_ALREADY_EXIST
  HARBOR_NOT_EXIST
  NO_PERMISSIONS
  INVALID_INPUT
}

input FairwayCardInput {
  id: String!
  version: String!
  operation: Operation!
  name: TextInput!
  n2000HeightSystem: Boolean!
  group: String!
  status: Status!
  fairwayIds: [Float!]!
  primaryFairwayId: [SelectedFairwayInput!]
  secondaryFairwayId: [SelectedFairwayInput!]
  additionalInfo: TextInput
  lineText: TextInput
  designSpeed: TextInput
  generalInfo: TextInput
  anchorage: TextInput
  navigationCondition: TextInput
  iceCondition: TextInput
  attention: TextInput
  speedLimit: TextInput
  visibility: TextInput
  vesselRecommendation: TextInput
  mareographs: [Int!]
  windRecommendation: TextInput
  trafficService: TrafficServiceInput
  harbors: [String!]
  pictures: [PictureInput!]
  pilotRoutes: [Int!]
  temporaryNotifications: [TemporaryNotificationInput!]
}

input PictureInput {
  id: String!
  sequenceNumber: Int
  orientation: Orientation!
  rotation: Float
  scaleWidth: String
  scaleLabel: String
  text: String
  lang: String
  groupId: Float
  modificationTimestamp: Float
  legendPosition: String
}

enum Orientation {
  PORTRAIT
  LANDSCAPE
}

input TrafficServiceInput {
  pilot: PilotInput
  vts: [VTSInput]
  tugs: [TugInput]
}

input TugInput {
  name: TextInput!
  phoneNumber: [String!]
  fax: String
  email: String
}

input VHFInput {
  name: TextInput
  channel: String!
}

input VTSInput {
  name: TextInput!
  phoneNumber: String
  email: [String]
  vhf: [VHFInput]
}

input PilotInput {
  email: String
  phoneNumber: String
  fax: String
  internet: String
  extraInfo: TextInput
  places: [PilotPlaceInput!]
}

input PilotPlaceInput {
  id: Float!
  pilotJourney: String
}

input HarborInput {
  id: String!
  version: String!
  n2000HeightSystem: Boolean!
  operation: Operation!
  status: Status!
  phoneNumber: [String]
  fax: String
  email: String
  internet: String
  name: TextInput!
  company: TextInput
  extraInfo: TextInput
  harborBasin: TextInput
  cargo: TextInput
  quays: [QuayInput]
  geometry: GeometryInput!
}

enum Status {
  DRAFT
  PUBLIC
  REMOVED
}

input GeometryInput {
  lat: String!
  lon: String!
}

input TextInput {
  fi: String!
  sv: String!
  en: String!
}

input QuayInput {
  name: TextInput
  length: String
  sections: [SectionInput]
  extraInfo: TextInput
  geometry: GeometryInput
}

input SectionInput {
  name: String
  depth: String
  geometry: GeometryInput
}

input TemporaryNotificationInput {
  content: TextInput!
  startDate: String!
  endDate: String
}

input SelectedFairwayInput {
  id: Int!
  sequenceNumber: Int!
}

type FairwayCardOrHarbor {
  id: String!
  version: String!
  type: ContentType!
  name: Text!
  group: String
  fairwayIds: [Int!]
  n2000HeightSystem: Boolean!
  status: Status!
  creator: String
  modifier: String
  modificationTimestamp: Float
  temporaryNotifications: [TemporaryNotification]
}

enum ContentType {
  CARD
  HARBOR
}

type CurrentUser {
  uid: String!
  name: String!
  roles: [String!]!
}

type FairwayCard {
  id: String!
  version: String!
  name: Text!
  n2000HeightSystem: Boolean!
  group: String
  status: Status
  creator: String
  creationTimestamp: Float
  modifier: String
  modificationTimestamp: Float
  fairwayIds: String
  fairways: [Fairway!]!
  additionalInfo: Text
  lineText: Text
  designSpeed: Text
  generalInfo: Text
  anchorage: Text
  navigationCondition: Text
  iceCondition: Text
  attention: Text
  speedLimit: Text
  visibility: Text
  vesselRecommendation: Text
  mareographs: [Mareograph!]
  windRecommendation: Text
  trafficService: TrafficService
  harbors: [Harbor!]
  pictures: [Picture!]
  pilotRoutes: [PilotRoute!]
  temporaryNotifications: [TemporaryNotification!]
}

type Picture {
  id: String!
  sequenceNumber: Int
  orientation: Orientation!
  rotation: Float
  text: String
  lang: String
  scaleLabel: String
  scaleWidth: String
  groupId: Float
  modificationTimestamp: Float
  legendPosition: String
}

type TrafficService {
  pilot: Pilot
  vts: [VTS]
  tugs: [Tug]
}

type Tug {
  name: Text
  phoneNumber: [String]
  fax: String
  email: String
}

type VHF {
  name: Text
  channel: Int
}

type VTS {
  name: Text
  phoneNumber: String
  email: [String]
  vhf: [VHF]
}

type Pilot {
  email: String
  phoneNumber: String
  fax: String
  internet: String
  extraInfo: Text
  places: [PilotPlace!]
}

type PilotPlace {
  id: Float!
  name: Text!
  pilotJourney: Float
  geometry: GeometryPoint!
}

type Harbor {
  id: String!
  version: String
  status: Status
  n2000HeightSystem: Boolean
  creator: String
  creationTimestamp: Float
  modifier: String
  modificationTimestamp: Float
  quays: [Quay]
  name: Text
  company: Text
  phoneNumber: [String]
  fax: String
  email: String
  internet: String
  extraInfo: Text
  harborBasin: Text
  cargo: Text
  geometry: GeometryPoint
}

type Quay {
  name: Text
  length: Float
  sections: [Section]
  extraInfo: Text
  geometry: GeometryPoint
}

type Section {
  name: String
  depth: Float
  geometry: GeometryPoint
}

type Fairway {
  id: Int!
  name: Text
  primary: Boolean
  primarySequenceNumber: Int
  secondary: Boolean
  secondarySequenceNumber: Int
  startText: String
  endText: String
  lightingCode: String
  lighting: Text
  typeCode: String
  type: Text
  owner: String
  area: Text
  sizing: Sizing
  classifications: [Classification!]
  sizingVessels: [SizingVessel!]
  navigationLines: [NavigationLine!]
  areas: [Area!]
  restrictionAreas: [RestrictionArea!]
  prohibitionAreas: [ProhibitionArea!]
  boardLines: [Boardline!]
  turningCircles: [TurningCircle!]
}

type Classification {
  type: String
  fairwayClassCode: String
  fairwayClass: Text
}

type AreaFairway {
  fairwayId: Int!
  status: Int
  line: Int
  sizingSpeed: Float
  sizingSpeed2: Float
  sequenceNumber: Int
}

type Area {
  id: Int!
  name: String
  depth: Float
  draft: Float
  referenceLevel: String
  n2000depth: Float
  n2000draft: Float
  n2000ReferenceLevel: String
  direction: Float
  journalNumber: String
  verificationDate: String
  owner: String
  additionalInformation: String
  typeCode: Int
  type: String
  notationCode: Float
  notation: String
  operationStatusCode: Float
  operationStatus: String
  operationTypeCode: Float
  operationType: String
  operationDirectionCode: Float
  operationDirection: String
  fairways: [AreaFairway!]
  geometry: GeometryPolygon
}

type RestrictionArea {
  id: Int!
  types: [RestrictionType!]
  value: Float
  presenter: String
  journalNumber: String
  verificationDate: String
  modificationDate: String
  startDate: String
  endDate: String
  status: String
  source: String
  location: String
  municipality: String
  exception: String
  geometry: GeometryPolygon
  fairways: [RestrictionFairway!]
}

type RestrictionType {
  code: String
  text: String
}

type RestrictionFairway {
  fairwayId: Int!
}

type ProhibitionArea {
  id: Int!
  extraInfo: Text
  fairway: ProhibitionFairway!
}

type ProhibitionFairway {
  fairwayId: Int!
  name: Text
}

type BoardlineFairway {
  fairwayId: Int!
}

type Boardline {
  id: Int!
  direction: Float
  geometry: GeometryLine
  fairways: [BoardlineFairway!]
}

type TurningCircleFairway {
  fairwayId: Int!
}

type TurningCircle {
  id: Int!
  geometry: GeometryLine
  diameter: Int
  fairways: [TurningCircleFairway!]
}

type SizingVessel {
  typeCode: String
  type: String
  length: Float
  width: Float
  draft: Float
  size: Float
  bodyFactor: Float
}

type Sizing {
  normalTurningCircle: Float
  minimumTurningCircle: Float
  normalWidth: Float
  minimumWidth: Float
  additionalInformation: String
  mareograph: String
  reserveWater: String
}

type NavigationLine {
  id: Int!
  depth: Float
  draft: Float
  referenceLevel: String
  n2000depth: Float
  n2000draft: Float
  n2000ReferenceLevel: String
  direction: Float
  length: Float
  owner: String
  verificationDate: String
  journalNumber: String
  additionalInformation: String
  type: String
  typeCode: String
  geometry: GeometryLine
  fairways: [NavigationLineFairway!]
}

type NavigationLineFairway {
  fairwayId: Int!
  status: Int
  line: Int
}

type SafetyEquipmentFault {
  id: Int!
  name: Text
  equipmentId: Int!
  typeCode: Int!
  type: Text
  recordTime: Float!
  geometry: GeometryPoint!
}

type GeometryPolygon {
  type: String
  coordinates: [[[Float]]]
}

type GeometryLine {
  type: String
  coordinates: [[Float]]
}

type GeometryPoint {
  type: String
  coordinates: [Float]
}

type Text {
  fi: String
  sv: String
  en: String
}

type MarineWarning {
  id: Int!
  number: Int!
  area: Text!
  type: Text!
  location: Text!
  description: Text!
  startDateTime: Float
  endDateTime: Float
  dateTime: Float!
  notifier: String!
  equipmentId: Float
  lineId: Float
  areaId: Float
}

type PilotRoute {
  id: Int!
  name: String
}

type TemporaryNotification {
  content: Text
  startDate: String
  endDate: String
}

type Mareograph {
  id: Int!
  name: String
  dateTime: Float
  waterLevel: Float
  n2000WaterLevel: Float
  calculated: Boolean
}
