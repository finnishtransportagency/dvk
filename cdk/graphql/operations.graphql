query findFairways {
  fairways {
    ...fragment FairwayFragment on Fairway {
      id
      nameFI
      nameSV
    }
  }
}

query findFairway($id: Int) {
  fairway(id: $id) {
    ...Fairway
  }
}
