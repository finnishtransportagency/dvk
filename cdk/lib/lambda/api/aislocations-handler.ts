import { ALBEvent, ALBResult } from 'aws-lambda';
import { log } from '../logger';
import { fetchLocations } from './ais';
import { handleAisCall } from '../util';

export const handler = async (event: ALBEvent): Promise<ALBResult> => {
  log.info({ event }, `aislocations()`);
  return handleAisCall('aislocations', fetchLocations, ['application/geo+json']);
};
