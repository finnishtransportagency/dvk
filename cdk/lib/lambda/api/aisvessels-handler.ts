import { ALBEvent, ALBResult } from 'aws-lambda';
import { log } from '../logger';
import { fetchVessels } from './ais';
import { handleAisCall } from '../util';

export const handler = async (event: ALBEvent): Promise<ALBResult> => {
  log.info({ event }, `aisvessels()`);
  return handleAisCall('aisvessels', fetchVessels, ['application/json']);
};
