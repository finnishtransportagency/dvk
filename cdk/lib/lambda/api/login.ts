import { getCloudFrontPrivateKey, getCloudFrontPublicKeyId, getCognitoUrl } from '../environment';
import { getSignedCookies } from '@aws-sdk/cloudfront-signer';
import { ALBEvent, ALBResult } from 'aws-lambda';
import { log } from '../logger';
import JWT, { JwtPayload } from 'jsonwebtoken';
import jwkToPem from 'jwk-to-pem';
import axios from 'axios';

const cloudFrontDnsName = process.env.CLOUDFRONT_DNSNAME;
let cachedKeys: Record<string, string>;

// Fetch JWK's from Cognito or cache
const getPublicKeys = async (issuerUrl: string) => {
  if (!cachedKeys) {
    cachedKeys = {};
    const publicKeys = await axios.get(issuerUrl + '/.well-known/jwks.json');
    for (const key of publicKeys.data.keys) {
      cachedKeys[key.kid] = jwkToPem(key);
    }
    return cachedKeys;
  } else {
    return cachedKeys;
  }
};

export const validateJwtToken = async (token: string | undefined, dataToken: string): Promise<JwtPayload | undefined> => {
  if (!token) {
    log.debug('IAM JWT Token missing');
    return;
  }
  // Split token into parts
  const tokenParts = token.split('.');
  if (tokenParts.length < 2) {
    log.error('Invalid token');
    return;
  }

  // Parse header & payload from token parts
  const tokenHeader = JSON.parse(Buffer.from(tokenParts[0], 'base64').toString('utf-8'));
  const tokenPayload = JSON.parse(Buffer.from(tokenParts[1], 'base64').toString('utf-8'));

  // Fetch public keys from Cognito
  const publicKeys = await getPublicKeys(tokenPayload.iss);
  const publicKey = publicKeys[tokenHeader.kid];
  if (!publicKey) {
    log.debug('Public key not found');
    return;
  }

  // Verify token
  const result = JWT.verify(token, publicKey, { issuer: await getCognitoUrl() }) as JwtPayload;
  if (!result) {
    log.error('Failed to verify JWT');
    return;
  }

  // Check use access
  if (result.token_use !== 'access') {
    log.error('Invalid token use');
    return;
  }

  // Return decoded data token
  return JWT.decode(dataToken.replace(/=/g, '')) as JwtPayload;
};

export const handler = async (event: ALBEvent): Promise<ALBResult> => {
  try {
    let jwtToken;
    if (event.multiValueHeaders) {
      log.info({ headers: event.multiValueHeaders }, 'Request headers');
      const token = event.multiValueHeaders['x-iam-accesstoken'];
      const data = event.multiValueHeaders['x-iam-data'];
      if (token && data) {
        jwtToken = await validateJwtToken(token[0], data[0]);
      }
    }
    log.info('JwtToken: %o', jwtToken);
    // TODO: comment out once token received from ALB
    /*if (!jwtToken) {
      return {
        statusCode: 403,
      };
    }*/
    const cloudFrontPolicy = JSON.stringify({
      Statement: [
        {
          Resource: `https://${cloudFrontDnsName}/*`,
          Condition: {
            DateLessThan: {
              'AWS:EpochTime': Math.floor(new Date().getTime() / 1000) + 3600,
            },
          },
        },
      ],
    });

    const cookie = getSignedCookies({
      url: cloudFrontDnsName as string,
      privateKey: await getCloudFrontPrivateKey(),
      keyPairId: await getCloudFrontPublicKeyId(),
      policy: cloudFrontPolicy,
    });

    const setCookieAttributes = `; Domain=${cloudFrontDnsName}; Path=/; Secure; HttpOnly; SameSite=Lax`;

    const response: ALBResult = {
      statusCode: 302,
      multiValueHeaders: {
        Location: [`https://${cloudFrontDnsName}/yllapito/index.html`],
        'Set-Cookie': [
          `CloudFront-Key-Pair-Id=${cookie['CloudFront-Key-Pair-Id']}${setCookieAttributes}`,
          `CloudFront-Policy=${cookie['CloudFront-Policy']}${setCookieAttributes}`,
          `CloudFront-Signature=${cookie['CloudFront-Signature']}${setCookieAttributes}`,
        ],
      },
    };
    log.debug(response, 'Login response');
    return response;
  } catch (e) {
    log.error('Getting signed cookies failed: %s', e);
    return { statusCode: 500, body: String(e) };
  }
};
