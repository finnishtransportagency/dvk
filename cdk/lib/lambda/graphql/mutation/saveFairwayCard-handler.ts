import { AppSyncResolverEvent, AppSyncResolverHandler } from 'aws-lambda/trigger/appsync-resolver';
import {
  FairwayCard,
  FairwayCardInput,
  InputMaybe,
  Maybe,
  MutationSaveFairwayCardArgs,
  Operation,
  OperationError,
  PictureInput,
  Status,
} from '../../../../graphql/generated';
import { auditLog, log } from '../../logger';
import FairwayCardDBModel, { Picture } from '../../db/fairwayCardDBModel';
import {
  getPilotPlaceMap,
  getPilotRoutes,
  mapEmail,
  mapEmails,
  mapFairwayCardDBModelToGraphqlType,
  mapId,
  mapIds,
  mapInternetAddress,
  mapMandatoryText,
  mapPhoneNumber,
  mapPhoneNumbers,
  mapPilotJourney,
  mapString,
  mapText,
  mapVhfChannel,
} from '../../db/modelMapper';
import { CurrentUser, getCurrentUser } from '../../api/login';
import { diff } from 'deep-object-diff';
import { ConditionalCheckFailedException } from '@aws-sdk/client-dynamodb';
import { getExpires, getNewStaticBucketName } from '../../environment';
import { PutObjectTaggingCommand, S3Client } from '@aws-sdk/client-s3';
import { deleteCacheObjects } from '../cache';

// Feature cache s3 objects related to fairway card
const CACHE_KEYS = [
  'pilot',
  'harbor',
  'area1,2',
  'area3,4,5,6',
  'specialarea151,2,3,4,5,6',
  'specialarea21,2,3,4,5,6',
  'line1,2',
  'line3,4,5,6',
  'safetyequipment1,2,99',
  'safetyequipmentfault',
];

export function mapFairwayCardToModel(card: FairwayCardInput, old: FairwayCardDBModel | undefined, user: CurrentUser): FairwayCardDBModel {
  return {
    id: mapId(card.id),
    name: mapMandatoryText(card.name),
    status: card.status,
    n2000HeightSystem: !!card.n2000HeightSystem,
    group: mapString(card.group),
    creationTimestamp: old ? old.creationTimestamp : Date.now(),
    creator: old ? old.creator : `${user.firstName} ${user.lastName}`,
    modifier: `${user.firstName} ${user.lastName}`,
    modificationTimestamp: Date.now(),
    fairways: card.fairwayIds.map((id, idx) => {
      return {
        id,
        primary: card.primaryFairwayId ? card.primaryFairwayId === id : idx === 0,
        secondary: card.secondaryFairwayId ? id === card.secondaryFairwayId : idx === 0,
      };
    }),
    generalInfo: mapText(card.generalInfo),
    anchorage: mapText(card.anchorage),
    iceCondition: mapText(card.iceCondition),
    attention: mapText(card.attention),
    additionalInfo: mapText(card.additionalInfo),
    lineText: mapText(card.lineText),
    designSpeed: mapText(card.designSpeed),
    navigationCondition: mapText(card.navigationCondition),
    windRecommendation: mapText(card.windRecommendation),
    windGauge: mapText(card.windGauge),
    visibility: mapText(card.visibility),
    seaLevel: mapText(card.seaLevel),
    speedLimit: mapText(card.speedLimit),
    vesselRecommendation: mapText(card.vesselRecommendation),
    trafficService: {
      pilot: {
        email: mapEmail(card.trafficService?.pilot?.email),
        extraInfo: mapText(card.trafficService?.pilot?.extraInfo),
        fax: mapPhoneNumber(card.trafficService?.pilot?.fax),
        internet: mapInternetAddress(card.trafficService?.pilot?.internet),
        phoneNumber: mapPhoneNumber(card.trafficService?.pilot?.phoneNumber),
        places:
          card.trafficService?.pilot?.places?.map((p) => {
            return {
              id: p.id,
              pilotJourney: mapPilotJourney(p.pilotJourney),
            };
          }) ?? [],
      },
      tugs:
        card.trafficService?.tugs?.map((t) => {
          return {
            email: mapEmail(t?.email),
            fax: mapPhoneNumber(t?.fax),
            name: mapText(t?.name),
            phoneNumber: mapPhoneNumbers(t?.phoneNumber),
          };
        }) ?? null,
      vts:
        card.trafficService?.vts?.map((v) => {
          return {
            email: mapEmails(v?.email),
            name: mapText(v?.name),
            phoneNumber: mapPhoneNumber(v?.phoneNumber),
            vhf: v?.vhf?.map((v2) => {
              return {
                name: mapText(v2?.name),
                channel: mapVhfChannel(v2?.channel),
              };
            }),
          };
        }) ?? null,
    },
    harbors:
      card.harbors?.map((id) => {
        return { id };
      }) ?? null,
    pilotRoutes:
      card.pilotRoutes?.map((id) => {
        return { id };
      }) ?? null,
    fairwayIds: mapIds(card.fairwayIds),
    expires: card.status === Status.Removed ? getExpires() : null,
    pictures:
      card.pictures?.map((p) => {
        return {
          id: p.id,
          sequenceNumber: p.sequenceNumber ?? null,
          orientation: p.orientation,
          rotation: p.rotation ?? null,
          modificationTimestamp: p.modificationTimestamp ?? null,
          scaleLabel: mapString(p.scaleLabel, 20),
          scaleWidth: mapString(p.scaleWidth, 20),
          text: mapString(p.text, 100),
          lang: mapString(p.lang, 2),
          groupId: p.groupId ?? null,
          legendPosition: p.legendPosition ?? null,
        };
      }) ?? null,
  };
}

const s3Client = new S3Client({ region: 'eu-west-1' });

async function tagPictures(cardId: string, pictures: InputMaybe<PictureInput[]> | undefined, oldPictures: Maybe<Picture[]> | undefined) {
  const promises = [];
  const bucketName = getNewStaticBucketName();
  for (const picture of pictures ?? []) {
    const command = new PutObjectTaggingCommand({
      Key: `${cardId}/${picture.id}`,
      Bucket: bucketName,
      Tagging: { TagSet: [{ Key: 'InUse', Value: 'true' }] },
    });
    log.debug(`setting InUse to true for ${picture.id}`);
    promises.push(s3Client.send(command));
  }
  for (const oldPicture of oldPictures ?? []) {
    if (!pictures?.find((p) => p.id === oldPicture.id)) {
      const command = new PutObjectTaggingCommand({
        Key: `${cardId}/${oldPicture.id}`,
        Bucket: bucketName,
        Tagging: { TagSet: [{ Key: 'InUse', Value: 'false' }] },
      });
      promises.push(s3Client.send(command));
      log.debug(`setting InUse to false for ${oldPicture.id}`);
    }
  }
  await Promise.all(promises);
}

export const handler: AppSyncResolverHandler<MutationSaveFairwayCardArgs, FairwayCard> = async (
  event: AppSyncResolverEvent<MutationSaveFairwayCardArgs>
): Promise<FairwayCard> => {
  const user = await getCurrentUser(event);
  log.info(`saveFairwayCard(${event.arguments.card.id}, ${user.uid})`);
  const dbModel = await FairwayCardDBModel.get(event.arguments.card.id);
  const newModel = mapFairwayCardToModel(event.arguments.card, dbModel, user);
  log.debug('card: %o', newModel);
  await tagPictures(event.arguments.card.id, event.arguments.card.pictures, dbModel?.pictures);
  try {
    if (dbModel?.status === Status.Public && newModel.status !== Status.Public) {
      // Clear feature cache s3 objects related to fairway card
      await deleteCacheObjects(CACHE_KEYS);
    }
    await FairwayCardDBModel.save(newModel, event.arguments.card.operation);
  } catch (e) {
    if (e instanceof ConditionalCheckFailedException && e.name === 'ConditionalCheckFailedException') {
      throw new Error(event.arguments.card.operation === Operation.Create ? OperationError.CardAlreadyExist : OperationError.CardNotExist);
    }
    throw e;
  }
  if (event.arguments.card.operation === Operation.Update) {
    const changes = dbModel ? diff(dbModel, newModel) : null;
    auditLog.info({ changes, card: newModel, user: user.uid }, 'FairwayCard updated');
  } else {
    auditLog.info({ card: newModel, user: user.uid }, 'FairwayCard added');
  }
  const pilotMap = await getPilotPlaceMap();
  const pilotRoutes = await getPilotRoutes();
  return mapFairwayCardDBModelToGraphqlType(newModel, pilotMap, user, pilotRoutes);
};
