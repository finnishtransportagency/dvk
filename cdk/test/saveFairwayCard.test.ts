import { FairwayCardInput, Operation, OperationError, Status } from '../graphql/generated';
import { CurrentUser } from '../lib/lambda/api/login';
import { mapFairwayCardToModel } from '../lib/lambda/graphql/mutation/saveFairwayCard-handler';

const currentUser: CurrentUser = { firstName: 'Test', lastName: 'Test', uid: 'test', roles: [] };
test('test mapFairwayCardToModel minimal', () => {
  expect(() => mapFairwayCardToModel({} as FairwayCardInput, undefined, currentUser)).toThrow(OperationError.InvalidInput);
  const input: FairwayCardInput = {
    id: 'a1',
    name: { fi: 'test', sv: 'testsv', en: 'testen' },
    n2000HeightSystem: false,
    operation: Operation.Create,
    status: Status.Draft,
    group: '1',
    fairwayIds: [1],
  };
  expect(mapFairwayCardToModel(input, undefined, currentUser)).toMatchSnapshot({
    modificationTimestamp: expect.any(Number),
    creationTimestamp: expect.any(Number),
  });
});

test('test mapFairwayCardToModel all', () => {
  const input: FairwayCardInput = {
    id: 'a2',
    name: { fi: 'test2', sv: 'testsv2', en: 'testen2' },
    n2000HeightSystem: true,
    operation: Operation.Update,
    status: Status.Public,
    group: '1',
    fairwayIds: [1],
    anchorage: {
      fi: 'anchorage',
      sv: 'anchoragesv',
      en: 'anchorageen',
    },
    attention: {
      fi: 'attentionfi',
      sv: 'attentionsv',
      en: 'attentionen',
    },
    designSpeed: {
      fi: 'designSpeed',
      sv: 'designSpeedsv',
      en: 'designSpeeden',
    },
    generalInfo: {
      fi: 'generalInfo',
      sv: 'generalInfosv',
      en: 'generalInfoen',
    },
    harbors: ['1'],
    iceCondition: {
      fi: 'iceCondition',
      sv: 'iceConditionsv',
      en: 'iceConditionen',
    },
    lineText: {
      fi: 'lineText',
      sv: 'lineTextsv',
      en: 'lineTexten',
    },
    navigationCondition: {
      fi: 'navigationCondition',
      sv: 'navigationConditionsv',
      en: 'navigationConditionen',
    },
    primaryFairwayId: 1,
    seaLevel: {
      fi: 'seaLevel',
      sv: 'seaLevelsv',
      en: 'seaLevelen',
    },
    secondaryFairwayId: 1,
    speedLimit: {
      fi: 'speedLimit',
      sv: 'speedLimitsv',
      en: 'speedLimiten',
    },
    trafficService: {
      pilot: {
        email: 'i@fi',
        extraInfo: {
          fi: 'extraInfo',
          sv: 'extraInfosv',
          en: 'extraInfoen',
        },
        fax: '123',
        internet: 'www',
        phoneNumber: '1234',
        places: [{ id: 1, pilotJourney: 0 }],
      },
      tugs: [
        {
          name: {
            fi: 'name',
            sv: 'namesv',
            en: 'nameen',
          },
          email: 'test@fi',
          fax: '12',
          phoneNumber: ['+358'],
        },
      ],
      vts: [
        {
          name: {
            fi: 'name',
            sv: 'namesv',
            en: 'nameen',
          },
          email: ['email@fi'],
          phoneNumber: '789',
          vhf: [
            {
              name: {
                fi: 'name',
                sv: 'namesv',
                en: 'nameen',
              },
              channel: 0,
            },
          ],
        },
      ],
    },
    vesselRecommendation: {
      fi: 'vesselRecommendation',
      sv: 'vesselRecommendationsv',
      en: 'vesselRecommendationen',
    },
    visibility: {
      fi: 'visibility',
      sv: 'visibilitysv',
      en: 'visibilityen',
    },
    windGauge: {
      fi: 'windGauge',
      sv: 'windGaugesv',
      en: 'windGaugeen',
    },
    windRecommendation: {
      fi: 'windRecommendation',
      sv: 'windRecommendationsv',
      en: 'windRecommendationen',
    },
  };
  expect(mapFairwayCardToModel(input, undefined, currentUser)).toMatchSnapshot({
    modificationTimestamp: expect.any(Number),
    creationTimestamp: expect.any(Number),
  });
});
