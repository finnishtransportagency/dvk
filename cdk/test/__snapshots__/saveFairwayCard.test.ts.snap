// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test mapFairwayCardToModel all 1`] = `
{
  "additionalInfo": null,
  "anchorage": {
    "en": "anchorageen",
    "fi": "anchorage",
    "sv": "anchoragesv",
  },
  "attention": {
    "en": "attentionen",
    "fi": "attentionfi",
    "sv": "attentionsv",
  },
  "creationTimestamp": Any<Number>,
  "creator": "Test Test",
  "designSpeed": {
    "en": "designSpeeden",
    "fi": "designSpeed",
    "sv": "designSpeedsv",
  },
  "expires": null,
  "fairwayIds": "#1#",
  "fairways": [
    {
      "id": 1,
      "primary": true,
      "secondary": true,
    },
  ],
  "generalInfo": {
    "en": "generalInfoen",
    "fi": "generalInfo",
    "sv": "generalInfosv",
  },
  "group": "1",
  "harbors": [
    {
      "id": "1",
    },
  ],
  "iceCondition": {
    "en": "iceConditionen",
    "fi": "iceCondition",
    "sv": "iceConditionsv",
  },
  "id": "a2",
  "lineText": {
    "en": "lineTexten",
    "fi": "lineText",
    "sv": "lineTextsv",
  },
  "mareographs": [
    {
      "id": 1,
    },
    {
      "id": 2,
    },
    {
      "id": 3,
    },
    {
      "id": 4,
    },
  ],
  "modificationTimestamp": Any<Number>,
  "modifier": "Test Test",
  "n2000HeightSystem": true,
  "name": {
    "en": "testen2",
    "fi": "test2",
    "sv": "testsv2",
  },
  "navigationCondition": {
    "en": "navigationConditionen",
    "fi": "navigationCondition",
    "sv": "navigationConditionsv",
  },
  "pictures": null,
  "pilotRoutes": null,
  "seaLevel": undefined,
  "speedLimit": {
    "en": "speedLimiten",
    "fi": "speedLimit",
    "sv": "speedLimitsv",
  },
  "status": "PUBLIC",
  "temporaryNotifications": null,
  "trafficService": {
    "pilot": {
      "email": "i@fi",
      "extraInfo": {
        "en": "extraInfoen",
        "fi": "extraInfo",
        "sv": "extraInfosv",
      },
      "fax": "12345",
      "internet": "www",
      "phoneNumber": "12345",
      "places": [
        {
          "id": 1,
          "pilotJourney": 0,
        },
      ],
    },
    "tugs": [
      {
        "email": "test@fi",
        "fax": "12222",
        "name": {
          "en": "nameen",
          "fi": "name",
          "sv": "namesv",
        },
        "phoneNumber": [
          "+35812",
        ],
      },
    ],
    "vts": [
      {
        "email": [
          "email@fi",
        ],
        "name": {
          "en": "nameen",
          "fi": "name",
          "sv": "namesv",
        },
        "phoneNumber": "78912",
        "vhf": [
          {
            "channel": 0,
            "name": {
              "en": "nameen",
              "fi": "name",
              "sv": "namesv",
            },
          },
        ],
      },
    ],
  },
  "version": "v0",
  "vesselRecommendation": {
    "en": "vesselRecommendationen",
    "fi": "vesselRecommendation",
    "sv": "vesselRecommendationsv",
  },
  "visibility": {
    "en": "visibilityen",
    "fi": "visibility",
    "sv": "visibilitysv",
  },
  "windGauge": {
    "en": "windGaugeen",
    "fi": "windGauge",
    "sv": "windGaugesv",
  },
  "windRecommendation": {
    "en": "windRecommendationen",
    "fi": "windRecommendation",
    "sv": "windRecommendationsv",
  },
}
`;

exports[`test mapFairwayCardToModel minimal 1`] = `
{
  "additionalInfo": null,
  "anchorage": null,
  "attention": null,
  "creationTimestamp": Any<Number>,
  "creator": "Test Test",
  "designSpeed": null,
  "expires": null,
  "fairwayIds": "#1#",
  "fairways": [
    {
      "id": 1,
      "primary": true,
      "secondary": true,
    },
  ],
  "generalInfo": null,
  "group": "1",
  "harbors": null,
  "iceCondition": null,
  "id": "a1",
  "lineText": null,
  "mareographs": undefined,
  "modificationTimestamp": Any<Number>,
  "modifier": "Test Test",
  "n2000HeightSystem": false,
  "name": {
    "en": "testen",
    "fi": "test",
    "sv": "testsv",
  },
  "navigationCondition": null,
  "pictures": null,
  "pilotRoutes": null,
  "seaLevel": undefined,
  "speedLimit": null,
  "status": "DRAFT",
  "temporaryNotifications": null,
  "trafficService": {
    "pilot": {
      "email": null,
      "extraInfo": null,
      "fax": null,
      "internet": null,
      "phoneNumber": null,
      "places": [],
    },
    "tugs": null,
    "vts": null,
  },
  "version": "v0",
  "vesselRecommendation": null,
  "visibility": null,
  "windGauge": null,
  "windRecommendation": null,
}
`;
